pkg/proxy/socket.go

```diff
@@ -0,0 +1,398 @@
+
+const (
+	fieldConn = "conn"
+	fieldSize = "size"
+)
+
+type proxySocket struct {
+	listener net.Listener
+	closing  chan struct{}
+}
+
+func listenSocket(address string, mark int) (*proxySocket, error) {
+	socket := &proxySocket{
+		closing: make(chan struct{}),
+	}
+
+	addr, err := net.ResolveTCPAddr("tcp", address)
+	if err != nil {
+		return nil, err
+	}
+
+	family := syscall.AF_INET
+	if addr.IP.To4() == nil {
+		family = syscall.AF_INET6
+	}
+
+	fd, err := syscall.Socket(family, syscall.SOCK_STREAM, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = syscall.SetsockoptInt(fd, syscall.SOL_SOCKET, syscall.SO_REUSEADDR, 1); err != nil {
+		return nil, fmt.Errorf("unable to set SO_REUSEADDR socket option: %s", err)
+	}
+
+	if mark != 0 {
+		setFdMark(fd, mark)
+	}
+
+	sockAddr, err := ipToSockaddr(family, addr.IP, addr.Port, addr.Zone)
+	if err != nil {
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	if err := syscall.Bind(fd, sockAddr); err != nil {
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	if err := syscall.Listen(fd, 128); err != nil {
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	f := os.NewFile(uintptr(fd), addr.String())
+	defer f.Close()
+
+	socket.listener, err = net.FileListener(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return socket, nil
+}
+
+// Accept calls Accept() on the listen socket of the proxy
+func (s *proxySocket) Accept() (*connectionPair, error) {
+	c, err := s.listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+
+	pair := newConnectionPair()
+	pair.rx.conn = c
+	pair.rx.startWriter()
+
+	return pair, nil
+}
+
+// Close closes the proxy socket and stops accepting new connections
+func (s *proxySocket) Close() {
+	close(s.closing)
+	s.listener.Close()
+}
+
+type socketQueue chan []byte
+
+type proxyConnection struct {
+	rx bool
+
+	// queueStopCount is > 0 if the queue has been stopped, it may only be
+	// read with queueStopped()
+	queueStopCount int32
+
+	conn  net.Conn
+	queue socketQueue
+
+	stopQueue chan struct{}
+
+	queueClose func()
+}
+
+func newProxyConnection(rx bool, queueClose func()) *proxyConnection {
+	return &proxyConnection{
+		rx:         rx,
+		queue:      make(socketQueue, socketQueueSize),
+		stopQueue:  make(chan struct{}),
+		queueClose: queueClose,
+	}
+}
+
+func (c *proxyConnection) queueStopped() bool {
+	return atomic.LoadInt32(&c.queueStopCount) > 0
+}
+
+func fmtAddress(a net.Addr) string {
+	if a == nil {
+		return "unknown"
+	}
+
+	return a.String()
+}
+
+func (c *proxyConnection) String() string {
+	if c.rx {
+		if c.conn == nil {
+			return "rx:closed"
+		}
+
+		return fmt.Sprintf("rx:%s->%s",
+			fmtAddress(c.conn.RemoteAddr()),
+			fmtAddress(c.conn.LocalAddr()))
+	}
+
+	if c.conn == nil {
+		return "tx:closed"
+	}
+
+	return fmt.Sprintf("tx:%s->%s",
+		fmtAddress(c.conn.LocalAddr()),
+		fmtAddress(c.conn.RemoteAddr()))
+}
+
+func (c *proxyConnection) startWriter() {
+	go c.socketWriter()
+}
+
+func (c *proxyConnection) socketWriter() {
+writingLoop:
+	for {
```

Comment:
I don't see a `return` on  this function, is this on purpose?

Reply:
return for what?

Reply:
A break. This is an infinite loop, the socketWriter will run forever, no?

Reply:
It runs until it is told to stop:
```
                select {
                case <-c.stopQueue:
                        atomic.AddInt32(&c.queueStopCount, 1)
                        break writingLoop
```

Reply:
but isn't the `break writingLoop` going to the jump `writingLoop` which is before that same for loop?

Reply:
@aanm it won't https://github.com/golang/go/issues/85